### PR TITLE
Fixes #302 Performance Optimization for the function replace_query

### LINKED ISSUE #302: Performance Optimization for the function replace_query
> replace_query() does a loop which is not necessary...
> 
> here is an optimized version of db_pgsql.php:
> 
>  @@ -1223,15 +1230,12 @@ } else { - $query = $this->write_query("SELECT {$main_field} FROM {$this->table_prefix}{$table}"); - - while($column = $this->fetch_array($query)) + $search_bit = "{$main_field} = '".$replacements[$main_field]."'"; + $query = $this->write_query("SELECT COUNT(".$main_field.") as count FROM {$this->table_prefix}{$table} WHERE {$search_bit} LIMIT 1"); + + if($this->fetch_field($query, "count") >= 1) { - if($column[$main_field] == $replacements[$main_field]) - { - $update = true; - break; - } + $update = true; } }

Original report: http://dev.mybb.com/issues/2174
Reporter: Can Özdemir
Status: New

### PR DESCRIPTION
**Pull request by JN-Jones**
Replaced the function with the modification mentioned in #302. Note that I haven't been able to test the new function on a board but based on the code it should work.

### CODE REVIEW

**Comment by euantorano** [comment]
This will definitely need some testing, but I'm away from home at the minute - can anybody else test this please?

**Comment by Ben-Conway** [comment]
@mybb/sqa can we get this tested?

**Comment by PaulBender** [comment]
It's not working for me, getting this SQL error:

`MyBB has experienced an internal SQL error and cannot continue.

SQL Error:
    0 - ERROR: syntax error at or near "=" LINE 2: SET fid=2, uid='1', dateline=1481233223 WHERE Array='' ^ 
Query:
    UPDATE mybb_forumsread SET fid=2, uid='1', dateline=1481233223 WHERE Array='' `

All I did was view a forum.

File: inc/db_pgsql.php [hunk 1]
```diff
@@ -1321,51 +1321,33 @@ function replace_query($table, $replacements=array(), $default_field="", $insert
...
 		{
-			if(is_array($main_field))
-			{
-				return $this->update_query($table, $replacements, $search_bit);
-			}
-			else
-			{
-				return $this->update_query($table, $replacements, "{$main_field}=".$this->quote_val($replacements[$main_field]));
-			}
+			return $this->update_query($table, $replacements, "{$main_field}=".$this->quote_val($replacements[$main_field]));
```
Comment (Shade-) [inline inc/db_pgsql.php]: `$main_field` is an array. Should be:

`return $this->update_query($table, $replacements, "{$main_field[0]}=".$this->quote_val($replacements[$main_field[0]]));`

**Comment by JN-Jones** [comment]
@PaulBender @Shade- Can we get this retested? Don't have a working pgsql copy myself atm.

File: inc/db_pgsql.php (hunk 1 above)
Comment (Shade-) [inline inc/db_pgsql.php]: Yep, didn't noticed it :)

**Comment by Shade-** [comment]
I don't have it either, just debugged directly looking at the code. Ball passed to @PaulBender.

**Comment by PaulBender** [comment]
Looks good.

**Pull request merged**